zebra-network/src/peer_set/candidate_set/tests/vectors.rs

```diff
@@ -115,6 +130,77 @@ fn rejects_all_addresses_if_applying_offset_causes_an_underflow() {
     assert!(validated_peers.next().is_none());
 }
 
+/// Test that calls to [`CandidateSet::update`] are rate limited.
+#[test]
+fn candidate_set_updates_are_rate_limited() {
+    let runtime = Runtime::new().expect("Failed to create Tokio runtime");
+    let _guard = runtime.enter();
+
+    let address_book = AddressBook::new(&Config::default(), Span::none());
+    let mut candidate_set =
+        CandidateSet::new(Arc::new(Mutex::new(address_book)), mock_peer_service());
+
+    runtime.block_on(async move {
+        time::pause();
+
+        // Run the test for enough time for `update` to actually run three times
+        let time_limit =
+            Instant::now() + 3 * MIN_PEER_GET_ADDR_INTERVAL + StdDuration::from_secs(1);
+
+        while Instant::now() <= time_limit {
+            candidate_set
+                .update()
+                .await
+                .expect("Call to CandidateSet::update should not fail");
+
+            // Call `update` frequently enough to have at least two skipped calls
+            time::advance(MIN_PEER_GET_ADDR_INTERVAL / 3).await;
+        }
+    });
+}
+
+/// Test that a call to [`CandidateSet::update`] after a call to [`CandidateSet::update_inital`] is
+/// rate limited.
+#[test]
+fn candidate_set_update_after_update_initial_is_rate_limited() {
+    let runtime = Runtime::new().expect("Failed to create Tokio runtime");
+    let _guard = runtime.enter();
+
+    let address_book = AddressBook::new(&Config::default(), Span::none());
+    let mut candidate_set =
+        CandidateSet::new(Arc::new(Mutex::new(address_book)), mock_peer_service());
+
+    runtime.block_on(async move {
+        time::pause();
+
+        // Call `update_initial` first
+        candidate_set
+            .update_initial(GET_ADDR_FANOUT)
+            .await
+            .expect("Call to CandidateSet::update should not fail");
+
+        // The following two calls to `update` should be skipped
+        candidate_set
+            .update()
+            .await
+            .expect("Call to CandidateSet::update should not fail");
+        time::advance(MIN_PEER_GET_ADDR_INTERVAL / 2).await;
+        candidate_set
+            .update()
+            .await
+            .expect("Call to CandidateSet::update should not fail");
+
+        // After waiting for the minimum interval the call to `update` should succeed
```

Comment:
Strictly you should only need `MIN_PEER_GET_ADDR_INTERVAL/2 + 1` here, because it's the cumulative time that matters.
```suggestion
        // After waiting for at least the minimum interval the call to `update` should succeed
```